sklearn/linear_model/sag.py

```diff
@@ -0,0 +1,354 @@
+"""Solvers for Ridge and LogisticRegression using SAG algorithm"""
+
+# Authors: Danny Sullivan <dbsullivan23@gmail.com>
+#          Tom Dupre la Tour <tom.dupre-la-tour@m4x.org>
+#
+# Licence: BSD 3 clause
+
+import numpy as np
+import scipy.sparse as sp
+import warnings
+
+from ..utils import ConvergenceWarning
+from ..utils import check_array
+from ..utils.seq_dataset import ArrayDataset, CSRDataset
+from .sgd_fast import Log, SquaredLoss
+from .sag_fast import sag, get_auto_eta
+
+from .stochastic_gradient import SPARSE_INTERCEPT_DECAY
+
+
+def make_dataset(X, y, sample_weight, random_state):
+    # check which type of Sequential Dataset is needed
+    seed = random_state.randint(1, np.iinfo(np.int32).max)
```

Comment:
1 will be twice more likely than any other value instead of being uniform. I don't think this is that of a problem a problem in practice. Anyway we still need to put inline comments to make it explicit that `out_rand_r` should never be seeded with 0 and make an audit of our code to prevent that bug from happening.